src/iframe-helper.js

```diff
@@ -100,24 +103,36 @@ function getOrCreateListenForEvents(parentWin, iframe) {
 /**
  * Returns an mapping of event names to listenFor listeners.
  * @param {!Window} parentWin the window that created the iframe
- * @param {string} origin the child window's origin
+ * @param {string} sentinel the sentinel of the message
+ * @param {string} origin the source window's origin
  * @param {!Window} triggerWin the window that triggered the event
- * @return {?Object<string, !Array<function(!object)>>}
+ * @return {?Object<string, !Array<function(!Object, !Window, string)>>}
  */
-function getListenForEvents(parentWin, origin, triggerWin) {
-  const listenOrigin = getListenForOrigin(parentWin, origin);
+function getListenForEvents(parentWin, sentinel, origin, triggerWin) {
+  const listenSentinel = getListenForSentinel(parentWin, sentinel);
 
-  if (!listenOrigin) {
-    return listenOrigin;
+  if (!listenSentinel) {
+    return listenSentinel;
   }
 
+  // Find the entry for the frame.
+  // TODO(@nekodo): Add a WeakMap<Window, WindowEventsDef> cache to
+  //     speed up this process.
   let windowEvents;
-  for (let i = 0; i < listenOrigin.length; i++) {
-    const we = listenOrigin[i];
+  for (let i = 0; i < listenSentinel.length; i++) {
+    const we = listenSentinel[i];
     const contentWindow = we.frame.contentWindow;
     if (!contentWindow) {
-      setTimeout(dropListenOrigin, 0, listenOrigin);
-    } else if (contentWindow === triggerWin) {
+      setTimeout(dropListenSentinel, 0, listenSentinel);
+    } else if (sentinel === 'amp') {
+      // A non-3P code path, origin must match.
+      if (we.origin === origin && contentWindow == triggerWin) {
```

Comment:
The origin check is unnecessary here because `contentWindow === triggerWin` will guarantee it. I think that means we can remove `origin` entirely from this `listenEvents` code.